Optimization of the Embedding and Embedding-Bag CUDA Kernel

madsbk opened the pull request:
Re-implementation of the `embedding_dense_backward_cuda()` and the `embedding_bag_backward_cuda_sum_avg()` functions. 

#### Performance
Running a [Mortgage Workflow](https://github.com/EvenOldridge/MortgageWorkflowA) with a block size of 100K on a DXG-2 (single GPU), we see a 270% speedup:
```
Original version:    370,168 example/s
Optimized version: 1,034,228 example/s
```
The original version is bounded by the `EmbeddingBag_accGradParametersKernel_sum_avg`, which takes 70% of the CUDA execution time. In the optimized version, the optimized kernel now takes only 17% of the time. 

#### Greater Numerical Stability 
An added benefit is greater numerical stability. Instead of doing a flat sum where a single variable are used to accumulate the weights, this code uses two-steps where each GPU-thread computes a sub-result defined by `NROWS_PER_THREAD` before the final result are accumulated.

madsbk committed:
Simplified EmbeddingBag_accGradParametersKernel_sum_avg

madsbk committed:
Now spawning a thread per segment (not row)

madsbk committed:
Swapped the x- and y-axis

madsbk committed:
Added a scatter kernel

madsbk committed:
Added a common file for embedding CUDA kernels

madsbk committed:
Combined the bag and non-bag version into one function

madsbk committed:
Using ceil_div()

SsnL commented on test/test_nn.py:
```diff
@@ -61,7 +61,7 @@
 
 dtype2prec = {torch.float: 1e-5,
               torch.double: 1e-5,
-              torch.half: 1e-2}
```
didn't review other parts. but does this pr make the computation less precise?

madsbk replied:
Actually, it is more precise. 
But now that I have also implemented the non-bag version, the reduction in the comparison sensitivity isn't need.

madsbk committed:
Testing: reverted the reduced comparison precision

Since the non-bag version now also uses the optimized version,
this is not needed anymore.

madsbk committed:
added docs

mrshenli commented:
@pytorchbot retest this please

mrshenli commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+
+__global__
+void split_segment_sizes_kernel(int64_t *ret, const int64_t *segment_offsets,
+                                int64_t num_of_segments, int64_t blocksize, int64_t numel) {
```
My first intuition of seeing `blocksize` is that it refers the CUDA block size, but it is actually `NROWS_PER_THREAD`. Can we come up with a different name to avoid conflicting with CUDA block?

madsbk replied:
Agree, it is now `constexpr`

mrshenli commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+// The maximum block size in CUDA
+constexpr int MAX_BLOCK_SIZE = 1024;
+// Number of rows each thread should sum initially
+constexpr int NROWS_PER_THREAD = 10;
```
how does this number impact the performance?

madsbk replied:
Added some info here: https://github.com/pytorch/pytorch/pull/22016/commits/14a597baca9e0fe47a46a61e8627f5f2859872d2

mrshenli commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+        const int64_t *segment_sizes_offsets,
+        const int64_t *segment_offsets,
+        int64_t num_of_segments,
+        int64_t blocksize) {
```
ditto

mrshenli reviewed (commented):
The numbers look awesome! Given that this adds extra complexity to speed up cases with a lot of duplicated indices, will it be slower if there is no duplication at all?

mrshenli commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+  accscalar_t weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    const accscalar_t scale = count ? (accscalar_t)1.0 / count[idx] : 1.0;
+    const int gradOutputRow = indices[idx] * stride;
```
I might miss something. It seems to me `idx` computed based on sorted indices, but `indices` here is the original index array. Is it correct to use `indices[idx]` here?

madsbk replied:
I think it is correct, it is based on: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L146

mrshenli commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+  }
+  const int idx_begin = segment_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?numel:segment_offsets[id+1];
+  if (idx_begin == padding_idx) {
```
Isn't `padding_idx` defined based on the embedding index instead of the index of indices array? It seems to me `padding_idx` needs to be compared with `original_indices[x]` instead of `idx_begin`.

ngimel replied:
it is compared with indices[idx_begin] below, so this indeed seems wrong or deserves a comment

mrshenli commented:
@madsbk is there any existing test that checks if the results are correct? or do we need to add some? It's quite non-trivial (thanks for implementing it!!) and I am not confident that a review is sufficient to guarantee the correctness.

madsbk committed:
Using include brackets instead of quotes

madsbk committed:
now using NROWS_PER_THREAD directly in the kernels

madsbk committed:
reformat

madsbk committed:
Changed var names to make more sense and added a lot of comments

madsbk committed:
typo

madsbk committed:
padding_idx is now checked against `indices[idx]`

madsbk commented:
@mrshenli, `test_nn.py` has some tests of both bag and non-bag embeddings. I have been using `py.test test/test_nn.py -k Embedding` while developing. However, there is no test of the `padding_idx` argument in `test_nn.py`; it might be tested somewhere else?

madsbk committed:
clean up

madsbk commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,311 @@
+    split_segment_offsets_kernel<<<ceil_div(num_of_segments, 32), 32, 0, stream>>> (
+            thrust::raw_pointer_cast(split_segment_offsets.data()),
+            thrust::raw_pointer_cast(split_segment_sizes.data()),
+            thrust::raw_pointer_cast(split_segment_sizes_offsets.data()),
```
I guess we could compute the size in the kernel but now that we have it, we might as well use it?

madsbk commented:
@mrshenli, this PR introduces a minor overhead if there is no duplication at all. But I do not think it is worth it to have the old code still around for that special case. 
I suspect that the overhead of checking for the no-duplicates case is similar to the current overhead. However, it will reduce the memory usage since it avoid the use of the temporary tensor `grad_weight_per_segment`.

mrshenli reviewed (approved):
Thanks for contributing!

BTW, I saw there is a `test_embedding_padding_idx` in `test_nn.py` for non-bag embeddings.

ngimel commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,326 @@
+  // We split the segments up into sizes of `NROWS_PER_THREAD`
+  // Compute the number partial-segments per segment (some partial-segments 
+  // may not be the full `NROWS_PER_THREAD` number of rows)
+  thrust::device_vector<int64_t> partials_per_segment(num_of_segments);
```
thrust::device_vector with default allocator should never be used. It calls cudaMalloc on construction and cudaFree on destruction

madsbk replied:
Fixed in #22401

ngimel commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,326 @@
+          partials_per_segment_offset.begin());
+
+  // The total number of partial-segments is the sum of `partials_per_segment_offset`
+  const int num_of_partial_segments = partials_per_segment[num_of_segments-1] +
```
Direct accesses to device_vector result in cudaMemcopy, synchronizing no less. They should be used very sparingly, if there's literally no way around it.

ngimel commented on aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu:
```diff
@@ -0,0 +1,326 @@
+  accscalar_t weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    const int64_t target_row = indices[idx];
+    if (target_row != padding_idx) {
```
target_row is the same for all indices in the loop, thus this check should be moved to before the loop. I agree it does not affect correctness and performance, but it affects code readability.

madsbk replied:
Fixed in #22401

ngimel reviewed (commented):
For fp16, this truncates inermediate gradients per segment to fp16, which can hurt accuracy and lead to overflows/underflows. Also, this launches many more kernels than original implementation, which will likely hurt performance in cpu-bound cases. Finally, device_vector with default allocator should never be used.
